### PR TITLE
apm-em-plugins-nxdefs-tracking

### PR DESCRIPTION
**Pull request by mkuehbach**
This is a tracking branch to be used for creating specific development installations of the pynxtools-apm and pynxtools-em readers to enable a consolidation and testing of the correct NeXus definitions on the fairmat branch that needs to be negotiated when creating a pypi version for pynxtools which supports definitions that are working with all dependant pynxtools-plugin(s)

### CODE REVIEW

File: pynxtools/nexus/nexus.py [hunk 1]
```diff
@@ -1,6 +1,5 @@
 # pylint: disable=too-many-lines
-"""Read files from different format and print it in a standard NeXus format
-"""
+"""Read files from different format and print it in a standard NeXus format"""
```
Comment (domna) [inline pynxtools/nexus/nexus.py:2]: I think the cleanest and most readable way to have this docstrings is the following:

```suggestion
"""
Read files from different format and print it in a standard NeXus format
"""
```

This is also how nomad does it.

File: tests/eln_mapper/test_eln_mapper.py [hunk 2]
```diff
@@ -1,5 +1,4 @@
-"""This test is dedicated generate_eln converter tool.
-"""
+"""This test is dedicated generate_eln converter tool."""
```
Comment (domna) [inline tests/eln_mapper/test_eln_mapper.py:1]:
```suggestion
"""
This test is dedicated generate_eln converter tool.
"""
```

File: tests/nexus/test_nexus.py [hunk 3]
```diff
@@ -1,6 +1,4 @@
-"""This is a code that performs several tests on nexus tool
-
-"""
+"""This is a code that performs several tests on nexus tool"""
```
Comment (domna) [inline tests/nexus/test_nexus.py:1]:
```suggestion
"""
This is a code that performs several tests on nexus tool
"""
```

File: tests/nexus/test_version.py [hunk 4]
```diff
@@ -1,6 +1,7 @@
 """
 Tests the version retrieval for the nexus definitions submodule
 """
+
```
Comment (domna) [inline tests/nexus/test_version.py:4]:
```suggestion
```

File: pynxtools/__init__.py [hunk 5]
```diff
@@ -1,6 +1,5 @@
-"""init file
+"""init file"""
```
Comment (domna) [inline pynxtools/__init__.py:1]:
```suggestion
"""
init file
"""
```


Do we want to have this docstring? I would suggest to remove it if it's just saying `init file`

File: pynxtools/dataconverter/readers/ellips/reader.py [hunk 6]
```diff
@@ -439,9 +440,9 @@ def read(
                 # using a proper unit parsing logic
                 template[f"/ENTRY[entry]/plot/DATA[{key}]/@units"] = "degree"
                 if dindx == 0 and index == 0:
-                    template[
-                        f"/ENTRY[entry]/plot/DATA[{key}]/@long_name"
-                    ] = f"{plot_name} (degree)"
+                    template[f"/ENTRY[entry]/plot/DATA[{key}]/@long_name"] = (
```
Comment (domna) [inline pynxtools/dataconverter/readers/ellips/reader.py:443]: Please revert this formatting change. It's confusing because nothing has chanced here and it has nothing to do with this PR

File: pynxtools/dataconverter/readers/ellips/reader.py [hunk 7]
```diff
@@ -472,9 +473,9 @@ def read(
             template["/ENTRY[entry]/program_name/@version"] = version("pynxtools")
         except PackageNotFoundError:
             pass
-        template[
-            "/ENTRY[entry]/program_name/@url"
-        ] = "https://github.com/FAIRmat-NFDI/pynxtools"
+        template["/ENTRY[entry]/program_name/@url"] = (
```
Comment (domna) [inline pynxtools/dataconverter/readers/ellips/reader.py:476]: Please revert this change as well.

**Review by domna — CHANGES_REQUESTED** [review]

File: pynxtools/_build_wrapper.py [hunk 8]
```diff
@@ -2,6 +2,7 @@
 Build wrapper for setuptools to create a nexus-version.txt file
 containing the nexus definitions verison.
 """
+
```
[the remark this answers is not in the document]
Comment (mkuehbach) [reply pynxtools/_build_wrapper.py:5]: Was auto-introduced by ruff which I accept as is. On the current master's [503ccf6](https://github.com/FAIRmat-NFDI/pynxtools/commit/503ccf6a90e322cd116298d6e9ae78f467f55b95) pyproject.toml ruff is not version constraints thus like on this feature branch, in my fresh venv checked via pip list ruff is 0.3.4 (the latest)

File: pynxtools/dataconverter/convert.py [hunk 9]
```diff
@@ -16,6 +16,7 @@
 # limitations under the License.
 #
 """This script runs the conversion routine using a selected reader and write out a NeXus file."""
+
```
[the remark this answers is not in the document]
Comment (mkuehbach) [reply pynxtools/dataconverter/convert.py:19]: Like above, I checked again and ran ruff on the latest commit of this feature branch locally, no auto-change was performed by ruff

File: pynxtools/dataconverter/readers/base/reader.py [hunk 10]
```diff
@@ -16,6 +16,7 @@
 # limitations under the License.
 #
 """The abstract class off of which to implement readers."""
+
```
[the remark this answers is not in the document]
Comment (mkuehbach) [reply pynxtools/dataconverter/readers/base/reader.py:19]: Same story

File: pynxtools/dataconverter/readers/ellips/mock.py [hunk 11]
```diff
@@ -16,6 +16,7 @@
 #
 
 """A generic class for generating duplicate outputs for ellipsometry"""
+
```
[the remark this answers is not in the document]
Comment (mkuehbach) [reply pynxtools/dataconverter/readers/ellips/mock.py:19]: Same story

File: pynxtools/dataconverter/readers/ellips/reader.py [hunk 12]
```diff
@@ -16,6 +16,7 @@
 # limitations under the License.
 #
 """An example reader implementation for the DataConverter."""
+
```
[the remark this answers is not in the document]
Comment (mkuehbach) [reply pynxtools/dataconverter/readers/ellips/reader.py:19]: Same story as above

File: pynxtools/dataconverter/helpers.py [hunk 13]
```diff
@@ -63,6 +63,9 @@ def get_all_defined_required_children_for_elem(xml_element):
     """Gets all possible inherited required children for a given NXDL element"""
     list_of_children_to_add = set()
     for child in xml_element:
+        tag = remove_namespace_from_tag(child.tag)
+        if tag not in ("group", "field", "attribute"):
+            continue
```
[the remark this answers is not in the document]
Comment (mkuehbach) [reply pynxtools/dataconverter/helpers.py:68]: Reads to me like a typical logic for fast skipping of elements. At the same time though, yeah what else for an NXDL should it be if not a group, field or attribute but then a raise ValueError

File: pynxtools/dataconverter/helpers.py [hunk 14]
```diff
@@ -637,6 +640,8 @@ def validate_data_dict(template, data, nxdl_root: ET.Element):
 def remove_namespace_from_tag(tag):
     """Helper function to remove the namespace from an XML tag."""
 
+    if not isinstance(tag, str):
+        return ""
```
[the remark this answers is not in the document]
Comment (mkuehbach) [reply pynxtools/dataconverter/helpers.py:644]: Several interpretations of what should be best practice:
1.) remove_namespace_from_tag should not even be called with non-str arguments (in C/C++ compiler will complain in such cases)
2.) change in lines 643-644 we accept as is but then does the callee have a check logic to prepare for dealing with the case that this function returns "" @sanbrock

File: pynxtools/nexus/nxdl_utils.py
[the remark this answers is not in the document]
Comment (mkuehbach) [reply pynxtools/nexus/nxdl_utils.py:1]: Agree, this level of care we need all the time for definitions as pynxtools + its plugins is nothing without the a working nexus_definitions version.

File: pynxtools/nexus/nexus.py [hunk 15]
```diff
@@ -9,7 +8,7 @@
 import click
 import h5py
 
-from pynxtools.nexus.nxdl_utils import *  # pylint: disable=wildcard-import, unused-wildcard-import
+from pynxtools.definitions.dev_tools.utils.nxdl_utils import *  # pylint: disable=wildcard-import, unused-wildcard-import
```
[the remark this answers is not in the document]
Comment (mkuehbach) [reply pynxtools/nexus/nexus.py:11]: Changed

File: pynxtools/dataconverter/readers/example/reader.py [hunk 16]
```diff
@@ -130,9 +131,9 @@ def read(
         my_compression_dict = {"compress": "string not to be compressed"}
         template["/ENTRY[entry]/test_compression/not_to_compress"] = my_compression_dict
         my_compression_dict2 = {"compress": np.array([1, 2, 3, 4])}
-        template[
-            "/ENTRY[entry]/test_compression/compressed_data"
-        ] = my_compression_dict2
+        template["/ENTRY[entry]/test_compression/compressed_data"] = (
```
[the remark this answers is not in the document]
Comment (mkuehbach) [reply pynxtools/dataconverter/readers/example/reader.py:134]: Ruff auto-formatted

File: pynxtools/dataconverter/readers/json_map/reader.py [hunk 17]
```diff
@@ -16,6 +16,7 @@
 # limitations under the License.
 #
 """An example reader implementation for the DataConverter."""
+
```
[the remark this answers is not in the document]
Comment (mkuehbach) [reply pynxtools/dataconverter/readers/json_map/reader.py:19]: Ruff auto

File: pynxtools/dataconverter/readers/json_yml/reader.py [hunk 18]
```diff
@@ -16,6 +16,7 @@
 # limitations under the License.
 #
 """An example reader implementation for the DataConverter."""
+
```
[the remark this answers is not in the document]
Comment (mkuehbach) [reply pynxtools/dataconverter/readers/json_yml/reader.py:19]: Ruff auto

File: pynxtools/dataconverter/readers/example/reader.py [hunk 19]
```diff
@@ -101,9 +102,9 @@ def read(
                 f"{my_path}/xarray_saved_small_calibration.h5:/axes/ax2",
             ]
         }
-        template[
-            "/ENTRY[entry]/test_virtual_dataset/concatenate_datasets"
-        ] = my_datasets
+        template["/ENTRY[entry]/test_virtual_dataset/concatenate_datasets"] = (
```
[the remark this answers is not in the document]
Comment (mkuehbach) [reply pynxtools/dataconverter/readers/example/reader.py:105]: Ruff auto-change

File: pynxtools/dataconverter/readers/example/reader.py [hunk 20]
```diff
@@ -17,6 +17,7 @@
 
 #
 """An example reader implementation for the DataConverter."""
+
```
[the remark this answers is not in the document]
Comment (mkuehbach) [reply pynxtools/dataconverter/readers/example/reader.py:20]: Ruff auto

File: pynxtools/dataconverter/readers/ellips/reader.py (hunk 7 above)
Comment (mkuehbach) [reply pynxtools/dataconverter/readers/ellips/reader.py:476]: Ruff auto

File: pynxtools/dataconverter/readers/example/reader.py (hunk 20 above)
[the remark this answers is not in the document]
Comment (mkuehbach) [reply pynxtools/dataconverter/readers/example/reader.py:20]: Ruff auto

File: pynxtools/dataconverter/readers/ellips/reader.py (hunk 6 above)
Comment (mkuehbach) [reply pynxtools/dataconverter/readers/ellips/reader.py:443]: No, ruff auto formatting

File: pynxtools/__init__.py (hunk 5 above)
Comment (mkuehbach) [reply pynxtools/__init__.py:1]: No, agree, apache default statement like for all nomad code files, will edit

File: pynxtools/_build_wrapper.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (domna) [reply pynxtools/_build_wrapper.py:5]: Ok, I see. Yeah, we probably should version constrain ruff so we don't introduce unintended changes. Thanks for clarification

File: pynxtools/dataconverter/helpers.py (hunk 13 above)
[the remark this answers is not in the document]
Comment (domna) [reply pynxtools/dataconverter/helpers.py:68]: The thing is I'm just a bit confused by the changes. While it is fine for me to merge this if this enables your parsers to work again we might want to collect the broader problem in an issue so we won't forget dealing with it.

File: tests/nexus/test_nexus.py (hunk 3 above)
Comment (mkuehbach) [reply tests/nexus/test_nexus.py:1]: Ruff auto formatted

File: tests/nexus/test_version.py (hunk 4 above)
Comment (mkuehbach) [reply tests/nexus/test_version.py:4]: Ruff auto

File: tests/eln_mapper/test_eln_mapper.py (hunk 2 above)
Comment (mkuehbach) [reply tests/eln_mapper/test_eln_mapper.py:1]: Ruff auto

File: pynxtools/nexus/nexus.py (hunk 1 above)
Comment (mkuehbach) [reply pynxtools/nexus/nexus.py:2]: Now I see your point on this non-default ruff styling issue, I don't what is our consensus? @sanbrock others?

File: pynxtools/nexus/nxdl_utils.py
[the remark this answers is not in the document]
Comment (mkuehbach) [reply pynxtools/nexus/nxdl_utils.py:1]: The situation is more involved:
The development of nexus (utils) switched from lxml to etree and now back to lxml.

I removed the wildcard import and replaced it with the following explicit list of imports:

from pynxtools.definitions.dev_tools.utils.nxdl_utils import (
    get_hdf_info_parent,
    get_nxdl_child,
    get_node_concept_path,
    get_local_name_from_xml,
    write_doc_string,
    try_find_units,
    check_attr_name_nxdl,
    get_required_string,
    get_best_child,
    get_node_name,
    add_base_classes,
    ET,
    walk_elist,
    get_nx_class,
    try_find_default,
    other_attrs,
    get_nexus_definitions_path,
    NxdlAttributeNotFoundError,
)

With this though most of the tests failed specifically all those which used at least one of the following functions:

    get_node_at_nxdl_path,
    get_enums,
    get_inherited_nodes,
    get_nx_classes
)

So I added these as well. Now that issue of wildcard import has been solved so do the test still run.

But indeed we have now a mixture of using nexus (utils code) some of which uses etree (fun's in nexus.py from pynxtools) and lxml (from nxdl_utils via nexus_definitions). What I do not like with the explicit code now is that the nexus.py source code file import functionalities which vscode shows to not be used until they are called through running nexus functions in test.

I agree that it makes sense to give this nexus utils code a round of revisiting, maybe refactoring and also maybe more guarding logic to deal with unexpected cases. However, this PR is certainly not the right place to put this in.

It was also never meant to touch on nexus utils at all if the simple introduction entryID(NXentry) would not have failed. Point is using this among us agreed upon correct syntax of our code revealed cases in nexus functions that demanded worked, classical chain.

@sanbrock what is your opinion about the fix here?

File: pynxtools/dataconverter/helpers.py (hunk 13 above)
[the remark this answers is not in the document]
Comment (mkuehbach) [reply pynxtools/dataconverter/helpers.py:68]: Fully agree, see:
https://github.com/FAIRmat-NFDI/pynxtools/issues/302 and
https://github.com/FAIRmat-NFDI/nexus_definitions/issues/206
to not forget about it

**Comment by mkuehbach** [comment]
@domna please merge then the PR related to https://github.com/FAIRmat-NFDI/pynxtools/pull/301, I can then merge the master in again run ruff again and then the formatting is not pulling the master branch afterwards

File: pynxtools/nexus/nexus.py [hunk 21]
```diff
@@ -4,11 +4,37 @@
...
+    get_node_concept_path,
+    get_local_name_from_xml,
+    write_doc_string,
+    try_find_units,
+    check_attr_name_nxdl,
+    get_required_string,
+    get_best_child,
+    get_node_name,
+    add_base_classes,
+    ET,
```
[the remark this answers is not in the document]
Comment (mkuehbach) [reply pynxtools/nexus/nexus.py:24]: Good point lemme fix this quick

[the remark this answers is not in the document]
Comment (mkuehbach) [reply pynxtools/nexus/nexus.py]: Done

File: pynxtools/nexus/nxdl_utils.py
[the remark this answers is not in the document]
Comment (domna) [reply pynxtools/nexus/nxdl_utils.py:1]: Ah sorry I overlooked this comment in the last review round.
So we use certain functions just in the tests, right? In that case I would not import them in nexus but just import them directly from `dev_tools`. Generally, I think it would be best to directly import from `dev_tools` everywhere instead of this indirect import via `nexus.py`
